artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/ServerPacketDecoder.java

```diff
@@ -83,16 +85,34 @@
 
 public class ServerPacketDecoder extends ClientPacketDecoder {
 
+   private static final int UUID_LENGTH = 36;
+   private static final int DEFAULT_INTERNER_CAPACITY = 32;
    private static final long serialVersionUID = 3348673114388400766L;
-   public static final ServerPacketDecoder INSTANCE = new ServerPacketDecoder();
+   private SimpleString.Interner keysInterner;
+   private TypedProperties.StringValue.Interner valuesInterner;
 
-   private static SessionSendMessage decodeSessionSendMessage(final ActiveMQBuffer in, CoreRemotingConnection connection) {
+   public ServerPacketDecoder() {
+      this.keysInterner = null;
+      this.valuesInterner = null;
+   }
+
+   private void initializeInternersIfNeeded() {
```

Comment:
one interner per session :O ouch - an interner ideally should be shared within JVM so benefits can be reaped anywhere.

Reply:
It has a different usage than a JVM interner.
It Is more a "per-connection last recently use fixed size pool": I've chosen to call it interner just for brevity eheh

Reply:
If though you think about it address SimpleString really can be shared every where, if the aim is to achieve / avoid creation of SimpleString's on the heap where its the same address really, then this can be achieved now in java 8 easily with a normal interner and creating objects within functions.

Reply:
I'm not avoiding only that cost, but the decoding and copy too: that's why I'm not using byte[] but directly mapping ByteBuf to intern the keys/values.

Reply:
you're not making a saving of avoiding iterating though the bytes, to check if equal/hash you're still having to process through that bytebuf bringing the bytes (if off heap) onto heap.

All creating SimpleString with data does is simple sets the byte[] as data, no decoding.

Reply:
> to check if equal/hash you're still having to process through that bytebuf bringing the bytes (if off heap) onto heap and doing a calc on them.

That cost is reduced avoiding bounds checking and limiting the length of the byte stream and if compare very well vs byte[] copy + SimpleString allocation (measured and tested with JMH).

>All creating SimpleString with data does is simple sets the byte[] as data, no decoding.

It is missing a couple of parts: byte[] allocation and byte[] copy (from the ByteBuf).
It seems sttrange, but it is avoiding the byte[] that is reduced by far most of the garbage

Reply:
Ok, so thats simple to change, why don't simple make SimpleString hold data as ByteBuf not byte[].

This then also would save this, and save further.

Reply:
sadly we can't do it with ease for different reasons:
- most byte[] operations are optimized by the JVM to avoid most of the bounds checks, on ByteBuf I've just provided a PR on Netty to improve this thing
- I wouldn't change SimpleString that is used everywhere (literally) on Artemis
- the lifecycle of the original ByteBuf is not predicatable: sometimes are pooled sometimes not, offheap/onheap...not that simple honestly

Reply:
Ok, going back to previous note: 
" but it is by avoiding the byte[] that is reduced by far most of the garbage"

We can do this on stack though with lamda's :)

Reply:
It is fragile due to how escape analysis work: looking just at the code the array escapes partially so it is not always true that will live on the stack...
And considering that the intern could steal it...probably it wont's work...
I've fought many years against these weird things of the JVM dear Michael :(

Reply:
see my note on the other comment, we can avoid the copy and byte[] object on heap too with keeping that on stack

Reply:
intern only steals it on first occurance so only that one escapes the stack as thats the one held in the intern, all others that are the same as the interned so wouldn't escape it.

Reply:
This is not how escape analysis work sadly: not on the standard Oracle JVM at least. 
Maybe Graal does a better job here because if allows partial escaping (first occurrence and never after too) while packing instances on the stack.

Reply:
I still think a global interning solution will be more beneficial especially with the other protocols that don't use SimpleString, but more use String then convert to SimpleString to keep ICoreMessage interface, but this will have same impact (especially in multi protocol broker env like with clients on AMQP and CORE).